federation/executor.go

```diff
@@ -59,67 +80,53 @@ func fetchSchema(ctx context.Context, e ExecutorClient, metadata interface{}) (*
-		if err != nil {
-			return nil, oops.Wrapf(err, "fetching schema %s", server)
-		}
-
-		var iq introspectionQueryResult
-		if err := json.Unmarshal(schema, &iq); err != nil {
-			return nil, oops.Wrapf(err, "unmarshaling schema %s", server)
-		}
-
-		schemas[server] = &iq
-	}
-
-	types, err := convertSchema(schemas)
-	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
-	}
-
-	introspectionSchema := introspection.BareIntrospectionSchema(types.Schema)
-	introspectionServer := &Server{schema: introspectionSchema}
-
-	executors["introspection"] = &DirectExecutorClient{Client: introspectionServer}
-	schema, err := introspection.RunIntrospectionQuery(introspection.BareIntrospectionSchema(introspectionServer.schema))
+func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Executor, error) {
+	return NewCustomExecutor(ctx, executors, &CustomExecutorArgs{})
+}
 
-	var iq introspectionQueryResult
-	if err := json.Unmarshal(schema, &iq); err != nil {
-		return nil, oops.Wrapf(err, "unmarshaling introspection schema")
+func NewCustomExecutor(ctx context.Context, executors map[string]ExecutorClient, c *CustomExecutorArgs) (*Executor, error) {
+	if c.schemaSyncer == nil {
+		c.schemaSyncer = newIntrospectionSchemaSyncer(ctx, executors, c.optionalArgs)
 	}
-
-	schemas["introspection"] = &iq
-	types, err = convertSchema(schemas)
-	if err != nil {
-		return nil, oops.Wrapf(err, "converting schemas error")
+	if c.schemaSyncInternvalSecond == 0 {
+		c.schemaSyncInternvalSecond = minSchemaSyncInterval
 	}
-
-	flattener, err := newFlattener(types.Schema)
+	planner, err := c.schemaSyncer.UpdatePlanner(ctx, c.optionalArgs)
 	if err != nil {
-		return nil, oops.Wrapf(err, "flattening schemas error")
+		return nil, oops.Wrapf(err, "failed to load schema")
 	}
-
-	// The planner is aware of the merged schema and what executors
-	// know about what fields
-	planner := &Planner{
-		schema:    types,
-		flattener: flattener,
+	executor := &Executor{
+		Executors: executors,
+		syncer: &Syncer{
+			ticker:       time.NewTicker(time.Duration(c.schemaSyncInternvalSecond) * time.Second),
+			schemaSyncer: c.schemaSyncer,
+			plannerMu:    &sync.RWMutex{},
+			planner:      planner,
+		},
 	}
+	go executor.poll(ctx, c.optionalArgs)
+	return executor, nil
+}
 
-	return &Executor{
-		Executors: executors,
-		planner:   planner,
-	}, nil
+func (e *Executor) poll(ctx context.Context, optionalArgs interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
+	for {
+		select {
+		case <-e.syncer.ticker.C:
+			newPlanner, err := e.syncer.schemaSyncer.UpdatePlanner(ctx, optionalArgs)
+			if err == nil && newPlanner != nil {
+				e.setPlanner(newPlanner)
+			}
+		}
+	}
 }
 
 func (e *Executor) runOnService(ctx context.Context, service string, typName string, keys []interface{}, kind string, selectionSet *graphql.SelectionSet, optionalArgs interface{}) ([]interface{}, interface{}, error) {
```

Comment:
Do you think it is better to change this function to take a planner passed by the Execute function? 
My thought is we want to make sure the planner is consistent during the lifetime of a query.